lib/common/zstd_internal.h

```diff
@@ -56,6 +56,48 @@ extern "C" {
 #define CHECK_F(f) { size_t const errcod = f; if (ERR_isError(errcod)) return errcod; }  /* check and Forward error code */
 #define CHECK_E(f, e) { size_t const errcod = f; if (ERR_isError(errcod)) return ERROR(e); }  /* check and send Error code */
 
+/**
+ * Return the specified error if the condition evaluates to true.
+ *
+ * In debug modes, prints additional information.
+ */
+#define RETURN_ERROR_IF(cond, err, ...) \
+  if (cond) { \
+    RAWLOG(3, "%s:%d: ERROR!: check %s failed, returning %s", __FILE__, __LINE__, ZSTD_QUOTE(cond), ZSTD_QUOTE(ERROR(err))); \
+    RAWLOG(3, ": " __VA_ARGS__); \
+    RAWLOG(3, "\n"); \
+    return ERROR(err); \
+  }
+
+/**
+ * Unconditionally return the specified error.
+ *
+ * In debug modes, prints additional information.
+ */
+#define RETURN_ERROR(err, ...) \
+  do { \
+    RAWLOG(3, "%s:%d: ERROR!: unconditional check failed, returning %s", __FILE__, __LINE__, ZSTD_QUOTE(ERROR(err))); \
+    RAWLOG(3, ": " __VA_ARGS__); \
+    RAWLOG(3, "\n"); \
+    return ERROR(err); \
+  } while(0);
+
+/**
+ * If the provided expression evaluates to an error code, returns that error code.
+ *
+ * In debug modes, prints additional information.
+ */
+#define FORWARD_ERROR(err, ...) \
```

Comment:
`FORWARD_ERROR()` is equivalent to `CHECK_F()`,
but add support for traces.

Is there a reason to have 2 separated variants ?
Would it be reasonable to fold logging support into `CHECK_F()` ?

Reply:
Yeah, I think it makes sense to eventually replace the `CHECK_*` macros. The barriers to that are:

1. I've only converted the zstd decompression code. I'd need to also convert all the compression code. I'm happy to do that work, but I wanted to put this PR up first to make sure everyone's happy with the approach taken.
2. The huffman codebase also defines its own `CHECK_{E,F}` macros, I guess because it needs to be freestanding from the zstd codebase. Should I convert those too? Or leave them be?

Reply:
1. If the modifications are done directly within `CHECK_F`, it's unnecessary to change the rest of the code to take advantage of it. 
  - There is a point though that maybe `FORWARD_ERROR()` is more expressive than `CHECK_F()`. I'm fine with that. That would require to change all `CHECK_F()` invocations though.
  - In which case, I was wondering about the name `FORWARD_IF_ERROR()`, because `FORWARD_ERROR()` feels like a definitive statement. But it's not what's happening.

2. Yes, keep Huffman code separate for the time being. We can try to unify that later on.